File: java/lab5/src/console/CommandManager.java
package src.console;

import java.util.HashMap;
import java.util.Scanner;

import src.baseabstractions.Command;
import src.basecollection.CollectionManager;
import src.baseobjects.Product;
import src.commands.*;

/**
 * Manages the registration and execution of console commands.
 * Provides a command registry and handles user input processing.
 *
 * @param <K> the type of keys used in the collection, must be Comparable
 * @param <V> the type of values stored in the collection, must extend Product
 */
public class CommandManager<K extends Comparable<K>, V extends Product> {
    private boolean isWorking = true;
    private static HashMap<String, Command<?, ? extends Product>> commands = new HashMap<>();
    Scanner scanner;

    /**
     * Constructs a CommandManager with the specified collection manager and
     * scanner.
     * Initializes and registers all available commands.
     *
     * @param collectionManager the collection manager to operate on
     * @param scanner           the scanner for reading user input
     */
    public CommandManager(CollectionManager<K, V> collectionManager, Scanner scanner) {
        this.scanner = scanner;

        commands.put("help", new HelpCommand<K, V>(collectionManager));
        commands.put("info", new InfoCommand<K, V>(collectionManager));
        commands.put("show", new ShowCommand<K, V>(collectionManager));
        commands.put("insert", new InsertCommand<K, V>(collectionManager));
        commands.put("update", new UpdateCommand<K, V>(collectionManager));
        commands.put("remove_key", new RemoveKeyCommand<K, V>(collectionManager));
        commands.put("clear", new ClearCommand<K, V>(collectionManager));
        commands.put("save", new SaveCommand<K, V>(collectionManager));
        commands.put("exit", new ExitCommand<K, V>(collectionManager));
        commands.put("remove_lower", new RemoveLowerCommand<K, V>(collectionManager));
        commands.put("replace_if_greater", new ReplaceIfGreaterCommand<K, V>(collectionManager));
        commands.put("remove_greater_key", new RemoveGreaterKeyCommand<K, V>(collectionManager));
        commands.put("count_greater_than_price", new CountGreaterThanPriceCommand<K, V>(collectionManager));
        commands.put("print_unique_unit_of_measure", new PrintUniqueUnitOfMeasureCommand<K, V>(collectionManager));
        commands.put("print_field_descending_price", new PrintFieldDescendingPriceCommand<K, V>(collectionManager));
    }

    /**
     * Returns the map of registered commands.
     *
     * @return HashMap containing command names as keys and Command objects as
     *         values
     */
    public static HashMap<String, Command<?, ? extends Product>> getCommands() {
        return commands;
    }

    /**
     * Adds a new command to the registry.
     *
     * @param key   the command name
     * @param value the Command object to register
     * @param <K>   the key type of the collection
     * @param <V>   the value type of the collection
     */
    public static <K extends Comparable<K>, V extends Product> void addCommand(String key, Command<K, V> value) {
        commands.put(key, value);
    }

    /**
     * Removes a command from the registry.
     *
     * @param key the command name to remove
     */
    public static void removeCommand(String key) {
        commands.remove(key);
    }

    /**
     * Returns the working state of the command manager.
     *
     * @return true if the manager is working, false otherwise
     */
    public boolean getWork() {
        return this.isWorking;
    }

    /**
     * Processes a single user command.
     * Reads input from the scanner, parses it, and executes the corresponding
     * command.
     * Handles command not found errors and execution exceptions.
     */
    public void existCommand() {
        try {
            System.out.flush();
            System.out.println("Введите команду: ");
            String command = scanner.nextLine().trim().toLowerCase();
            String[] args = command.split(" ");
            if (commands.containsKey(args[0])) {
                try {
                    commands.get(args[0]).execute(args);
                } catch (IllegalArgumentException e) {
                    System.out.println("Что-то пошло не так. " + e.getMessage() + "Попробуйте еще раз.");
                }
            } else {
                System.out.println("Команда \"" + args[0] + "\" не найдена.");
            }
        } catch (Exception e) {
            System.out.println("Something went wrong. " + e.getMessage());
            this.isWorking = false;
            System.exit(0);
        }
    }
}

File: java/lab6/client/src/main/kotlin/console/CommandManager.kt
package console

import abstractions.AbstractReaderWriter
import abstractions.Command
import commands.ClearCommand
import commands.CountGreaterThanPriceCommand
import commands.ExecuteScriptCommand
import commands.ExitCommand
import commands.HelpCommand
import commands.InfoCommand
import commands.InsertCommand
import commands.PrintFieldDescendingPriceCommand
import commands.PrintUniqueUnitOfMeasureCommand
import commands.RemoveGreaterKeyCommand
import commands.RemoveKeyCommand
import commands.RemoveLowerCommand
import commands.ReplaceIfGreaterCommand
import commands.ShowCommand
import commands.UpdateCommand
import connection.Request
import connection.Response
import connection.TcpClient
import kotlinx.serialization.json.Json

/**
 * Registers all commands, drives the interactive read-execute loop, and coordinates
 * serialization and network communication.
 *
 * In script mode field value lines must start with '>'. If a form expects a field
 * value but the next line is a command (no leading '>'), a [CommandInterruptedException]
 * is thrown by [FileReaderWriter]. [runFromFileReaderWriter] catches it, prints an error,
 * and resumes execution from the interrupting command line.
 *
 * @property tcpClient the TCP client used to send requests to the server
 * @property readerWriter the primary I/O channel (console)
 */
class CommandManager(
    private val tcpClient: TcpClient,
    private val readerWriter: AbstractReaderWriter
) {
    private val json = Json { ignoreUnknownKeys = true }

    val commands: MutableMap<String, Command> = mutableMapOf()

    init {
        val form = ProductConsoleForm(readerWriter)

        commands["help"] = HelpCommand(commands)
        commands["exit"] = ExitCommand()
        commands["info"] = InfoCommand()
        commands["show"] = ShowCommand()
        commands["insert"] = InsertCommand(form)
        commands["update"] = UpdateCommand(form)
        commands["remove_key"] = RemoveKeyCommand()
        commands["clear"] = ClearCommand()
        commands["remove_lower"] = RemoveLowerCommand(form)
        commands["replace_if_greater"] = ReplaceIfGreaterCommand(form)
        commands["remove_greater_key"] = RemoveGreaterKeyCommand()
        commands["count_greater_than_price"] = CountGreaterThanPriceCommand()
        commands["print_unique_unit_of_measure"] = PrintUniqueUnitOfMeasureCommand()
        commands["print_field_descending_price"] = PrintFieldDescendingPriceCommand()
        commands["execute_script"] = ExecuteScriptCommand(this)
    }

    /**
     * Runs the interactive command loop, reading from [readerWriter] until EOF.
     */
    fun run() {
        while (true) {
            readerWriter.write("Enter the command:")
            val line = readerWriter.readLine()
            if (line.isEmpty()) continue
            processLine(line)
        }
    }

    /**
     * Runs commands from the provided [fileReaderWriter], used by [ExecuteScriptCommand].
     * Command lines are read via [FileReaderWriter.readCommandLine].
     * When a [CommandInterruptedException] surfaces from a form, the interrupting command
     * line is extracted and execution continues from it.
     *
     * @param fileReaderWriter the file-backed I/O channel to read commands from
     */
    fun runFromFileReaderWriter(fileReaderWriter: FileReaderWriter) {
        patchFormReaderWriter(fileReaderWriter)
        try {
            var nextLine: String? = fileReaderWriter.readCommandLine()
            while (nextLine != null) {
                val line = nextLine
                nextLine = null
                try {
                    processLine(line)
                } catch (e: CommandInterruptedException) {
                    val commandName = line.trim().split(Regex("\\s+")).firstOrNull() ?: line
                    println("Command \"$commandName\" was not executed: object input was incomplete.")
                    nextLine = e.commandLine
                    continue
                }
                nextLine = fileReaderWriter.readCommandLine()
            }
        } finally {
            patchFormReaderWriter(readerWriter)
        }
    }

    private fun patchFormReaderWriter(readerWriter: AbstractReaderWriter) {
        val form = ProductConsoleForm(readerWriter)
        commands["insert"] = InsertCommand(form)
        commands["update"] = UpdateCommand(form)
        commands["remove_lower"] = RemoveLowerCommand(form)
        commands["replace_if_greater"] = ReplaceIfGreaterCommand(form)
    }

    private fun processLine(line: String) {
        val tokens = line.trim().split(Regex("\\s+"))
        val name = tokens[0].lowercase()
        val args = tokens.toTypedArray()

        val command = commands[name]
        if (command == null) {
            readerWriter.write("The \"$name\" command was not found.")
            return
        }

        try {
            command.start(args)
        } catch (e: CommandInterruptedException) {
            throw e
        } catch (e: IllegalArgumentException) {
            readerWriter.write("Argument error: ${e.message}")
            return
        } catch (e: NumberFormatException) {
            readerWriter.write("Invalid number format: ${e.message}")
            return
        }

        if (!command.isShouldBeSent) return

        val payload = extractPayload(command)
        val request = Request(commandName = name, serializedCommand = payload)

        val response: Response = try {
            tcpClient.send(request)
        } catch (e: IllegalStateException) {
            readerWriter.write("Server unavailable: ${e.message}")
            return
        }

        if (response.errorMessage != null) {
            readerWriter.write("Server error: ${response.errorMessage}")
            return
        }

        applyResult(command, response.serializedResult)
        command.finish()
    }

    private fun extractPayload(command: Command): String = when (command) {
        is InsertCommand -> command.toPayload()
        is UpdateCommand -> command.toPayload()
        is RemoveKeyCommand -> command.toPayload()
        is RemoveLowerCommand -> command.toPayload()
        is ReplaceIfGreaterCommand -> command.toPayload()
        is RemoveGreaterKeyCommand -> command.toPayload()
        is CountGreaterThanPriceCommand -> command.toPayload()
        else -> "{}"
    }

    private fun applyResult(command: Command, resultJson: String) {
        when (command) {
            is InfoCommand -> command.applyResult(resultJson)
            is ShowCommand -> command.applyResult(resultJson)
            is InsertCommand -> command.applyResult(resultJson)
            is UpdateCommand -> command.applyResult(resultJson)
            is RemoveKeyCommand -> command.applyResult(resultJson)
            is RemoveLowerCommand -> command.applyResult(resultJson)
            is ReplaceIfGreaterCommand -> command.applyResult(resultJson)
            is RemoveGreaterKeyCommand -> command.applyResult(resultJson)
            is CountGreaterThanPriceCommand -> command.applyResult(resultJson)
            is PrintUniqueUnitOfMeasureCommand -> command.applyResult(resultJson)
            is PrintFieldDescendingPriceCommand -> command.applyResult(resultJson)
            else -> {}
        }
    }
}